atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java

```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(3, 5))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(4, 5))).hasSize(replicas);
+
+        // exhaustively check all ranges
+        for (int i = 0; i < partitions; i++) {
+            for (int j = i + 1; j <= partitions; j++) {
+                for (int k = 1; k <= replicas; k++) {
+                    List<TokenRange> ranges = tokenRanges.subList(i, j);
+                    assertThat(cassandra.getCassandraServers(ranges))
+                            .as("(%d, %d, %d) servers for ranges %s", i, j, k, ranges)
+                            .hasSize((j - i) * replicas);
+                }
             }
         }
     }
 
+    @Test
+    public void testComplexMultiDatacenterMultiRackRefreshTokenRanges() {
+        int replicas = 3;
+        int partitions = 5;
+        Set<CassandraServer> nodes = generateNodes(replicas, partitions);
+        assertThat(nodes).allSatisfy(node -> assertThat(node.reachableProxyIps())
+                .allSatisfy(address -> assertThat(address.isUnresolved()).isTrue()));
+        CassandraService cassandra = clientPoolWithServers(nodes);
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+        assertThat(tokenRanges).hasSize(partitions);
+        assertThat(cassandra.refreshTokenRangesAndGetServers(() -> tokenRanges))
+                .hasSize(nodes.size())
+                .satisfies(servers -> {
+                    assertThat(Collections2.transform(servers, CassandraServer::cassandraHostName))
+                            .containsExactlyInAnyOrderElementsOf(
+                                    Collections2.transform(nodes, CassandraServer::cassandraHostName));
+                    assertThat(Collections2.transform(servers, CassandraServer::reachableProxyIps))
+                            .allSatisfy(ips -> assertThat(ips).allSatisfy(address -> {
+                                assertThat(address).isNotNull();
+                                assertThat(address.isUnresolved()).isFalse();
+                            }));
+                });
+    }
+
+    @Test
+    public void testComplexMultiDatacenterMultiRackPartialRefreshTokenRanges() {
+        int replicas = 3;
+        int partitions = 5;
+        Set<CassandraServer> nodes = generateNodes(replicas, partitions);
+        assertThat(nodes).allSatisfy(node -> assertThat(node.reachableProxyIps())
+                .allSatisfy(address -> assertThat(address.isUnresolved()).isTrue()));
+        CassandraService cassandra = clientPoolWithServers(nodes);
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+        assertThat(tokenRanges).hasSize(partitions);
+        assertThat(cassandra.refreshTokenRangesAndGetServers(() -> tokenRanges.subList(0, 3)))
+                .hasSize(9)
+                .satisfies(servers -> {
```

Comment:
quite hard to parse these constants and why they were chosen

Reply:
will adjust to make this clearer